StandardFiniteFieldsOSCAR.jl

```diff
@@ -1,443 +1,443 @@
-                vec = vcat(map( a -> coords(a) * pK, v)...)
-
-            end
-        end
-        # Now p is the minimal polynomial over F
-        # pmat gives the primitive powers in the tower basis for the new extension
-    end
-
-    vname = "x" * string(d)
-    L, X = FiniteField(polynomial(F, p), vname)
-    L.is_standardfinitefield = true
-    L.primitivepowersintowerbasis = pmat
-
-    return L
-
-end
-
-
-function standardfinitefield(p::ZZRingElem,n)
-    if !isprime(p)
-        error()
-    end
-    F = GF(p)
-    if !F.is_standardprimefield
-        F.is_standardprimefield = true
-        F.standardextensions = Dict{ZZRingElem, FinField}(1 => F)
-        F.primitivepowersintowerbasis = identity_matrix(F, 1)
-    end
-    ext = F.standardextensions
-    if haskey(ext, n)
-        return ext[n]
-    end
-
-    nfactorization = factor(ZZ(n));
-    nfactors = sort([r for (r,e) in nfactorization]);
-    lastfactor = nfactors[end]
-    nK = div(n,lastfactor)
-    K = standardfinitefield(p, nK)
-
-    stn = steinitznumberforprimedegree(p, lastfactor, nfactorization[lastfactor])
-    n1 = lastfactor^(nfactorization[lastfactor]-1)
-    q1 = p^n1
-
-    # for each element y in this list, we want to
-    # 1. call EmbedSteinitz(p, n1, nK, y)
-    # 2. this should give a number, we want to use ElementSteinitzNumber to get an element of K.
-    l = digits(stn, base = BigInt(q1))
-    c = map(y -> elementfromsteinitznumber(K, embedsteinitz(p, n1, nK, y)), l)
-    b = elementfromsteinitznumber(K, p^( findfirst(x -> x == div(nK, n1), std_mon_degs(nK))-1))
-
-    L = _extensionwithtowerbasis(K, lastfactor, c, b)
-    ext[n] = L
-    return L
-end
-
-
-
-
-GAP.Packages.load("StandardFF")
-
-function MyPoly(p,n)
-    return defining_polynomial(standardfinitefield(p,n))
-end
-
-function GAPPoly(p,n)
-    poly = GAP.evalstr( "StandardFiniteField(" * string(p) * ", " * string(n) * ")!.DefiningPolynomial!.CoefficientsOfUnivariatePolynomial")
-    poly = polynomial(GF(p), map(x -> GF(p)(x), GAP.gap_to_julia(Vector{GAP.FFE}, poly)))
-end
-
-function compare_poly(p,n)
-  F = GF(p)
-  F.(collect(coefficients(MyPoly(p, n)))) == F.(collect(coefficients(GAPPoly(p, n))))
-end
-
-p = ZZ(3)
-S = Set([ n for n in 6:64 if !compare_poly(p,n) ])
-
-p = ZZ(5)
-for n in 100:150
-    println(n)
-    @time F = standardfinitefield(p,n)
-    @time FG = GAP.Globals.StandardFiniteField(5,n)
-end
+ using Oscar
+
+ # NOTE: These give missing features to OSCAR/Nemo that will likely be added in the near future.
+ function (k::Nemo.FpField)(a::Vector)
+   @assert length(a) == 1
+   return k(a[1])
+ end
+ function (k::FqPolyRepField)(a::Vector)
+   return k(polynomial(GF(ZZ(characteristic(k))), a))
+ end
+ function coords(x::FinFieldElem)
+     return absolute_coordinates(x)
+ end
+ function coords(x::T) where T <: Union{fpFieldElem, FpFieldElem}
+     return [x]
+ end
+
+
+ # NOTE: We piggyback on the _attrs (type Dict{Symbol, Any}) property to define:
```

Comment:
But why? Why aren't you simply using the attribute system as it was intended to be used?

Reply:
I admit, I was under the impression that Julia didn't have an attribute system, only fields and properties as a workaraound (so get/setproperty could be overwritten without affecting get/setfield). Do get/setattribute come from julia Base or from a package? I can't find any documentation.

Reply:
They come from AbstractAlgebra, see https://nemocas.github.io/AbstractAlgebra.jl/stable/misc/#Attributes

This is where `__attrs` comes from ;-).

Reply:
Aaaah I see, I found this detailed in the Oscar documentation. I will use this.